### PR TITLE
Add witness-to-key and ec-to-witness commands and tests.

### PR DESCRIPTION
**Pull request by thecodefactory**

### CODE REVIEW

**Comment by thecodefactory** [comment]
Requires https://github.com/libbitcoin/libbitcoin-system/pull/1118

**Comment by thecodefactory** [comment]
Examples:

```
$ bx seed | bx ec-new | bx ec-to-public | bx ec-to-witness
bc1q0dcfy68pgz8pl59g3vlxvmjlwws5rhxssv03xp

$ bx seed | bx ec-new | bx ec-to-public | bx ec-to-witness --prefix tb
tb1qt2n5djprfhc2nln9938q5kx64seaxlvdekejtp
```

```
$ bx witness-to-key tb1qr47dd36u96r0fjle36hdygdnp0v6pwfgqe6jxg
31923086d7a16098a89b512c541c0bdea1f249b2bb3e9817382f7fbc087d5062

$ bx fetch-history 31923086d7a16098a89b512c541c0bdea1f249b2bb3e9817382f7fbc087d5062
transfers
{
    transfer
    {
        received
        {
            hash 56f87210814c8baef7068454e517a70da2f2103fc3ac7f687e32a228dc80e115
            height 988972
            index 1
        }
        value 16507846636160024577
    }
    transfer
    {
        spent
        {
            hash d869f854e1f8788bcff294cc83b280942a8c728de71eb709a2c29d10bfe21b7c
            height 989649
            index 0
        }
        value 18446744073709551615
    }
}
```

The amount values across v4 are incorrect (haven't gotten to the bottom of that yet).

Address history otherwise corresponds to https://www.blockchain.com/btctest/address/tb1qr47dd36u96r0fjle36hdygdnp0v6pwfgqe6jxg

**Comment by evoskuil** [comment]
I think we should provide the support in `wrap-encode` and `wrap-decode` by detecting parameters by size or name.

File: test/commands/ec-to-witness.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "command.hpp"
+
+BX_USING_NAMESPACES()
+
+BOOST_AUTO_TEST_SUITE(offline)
+BOOST_AUTO_TEST_SUITE(ec_to_witness__invoke)
```
Comment (evoskuil) [inline test/commands/ec-to-witness.cpp]: Why not use vector declarations in this suite as well?

File: test/commands/witness-to-key.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+#define BX_WITNESS_TO_KEY_WITNESS2 "tb1qq82ajthl5mlm50h6x70esvxs7atp3vfnjwp8z5kjdepsjqqw3zcsj5rufw"
+
+// Expectations.
+#define BX_WITNESS_TO_KEY_KEY1 "31923086d7a16098a89b512c541c0bdea1f249b2bb3e9817382f7fbc087d5062\n"
+#define BX_WITNESS_TO_KEY_KEY2 "edd374bd7ae9fe2c2cb9f20adbc675724fbb20b6f40b995ad7c9deb617d09d7b\n"
+
+BOOST_AUTO_TEST_CASE(witness_to_key__invoke__witness1__okay_output)
+{
+    BX_DECLARE_CLIENT_COMMAND(bc::explorer::commands::witness_to_key);
+    command.set_witness_address_argument({ BX_WITNESS_TO_KEY_WITNESS1, bc::system::wallet::witness_address::encoding::testnet_p2wpkh });
```
Comment (evoskuil) [inline test/commands/witness-to-key.cpp]: Reduce the line length with `using bc`, `using bc::explorer` and using bc::system::wallet` unless there are namespace conflicts.

**Comment by thecodefactory** [comment]
Updated to match latest changes in https://github.com/libbitcoin/libbitcoin-system/pull/1118

File: src/commands/ec-to-witness.cpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+using namespace bc::system::wallet;
+
+console_result ec_to_witness::invoke(std::ostream& output, std::ostream& error)
+{
+    // Bound parameters.
+    const auto& point = get_ec_public_key_argument();
+    const auto prefix = get_prefix_argument();
+    const auto format = get_address_format_option();
+
+    output << witness_address(point, format, prefix).encoded() << std::endl;
```
Comment (evoskuil) [inline src/commands/ec-to-witness.cpp]: All that generated code for this one line :)).

File: src/config/address_format.cpp [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+namespace libbitcoin {
+namespace explorer {
+namespace config {
+
+using namespace bc::system;
+using namespace bc::system::wallet;
+using namespace po;
+
+static const auto address_format_p2wpkh = "p2wpkh";
+static const auto address_format_p2wsh = "p2wsh";
```
Comment (evoskuil) [inline src/config/address_format.cpp]: Acronyms :/

File: test/commands/ec-to-witness.cpp [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+struct test_vector
+{
+    std::string public_key;
+    std::string prefix;
+    witness_address::address_format format;
+    std::string output;
+};
+
+const std::vector<test_vector> test_list =
```
Comment (evoskuil) [inline test/commands/ec-to-witness.cpp]: `test_vectors`?

File: test/commands/witness-to-key.cpp [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+using namespace bc::system::wallet;
+
+struct test_item
+{
+    std::string address;
+    witness_address::address_format format;
+    std::string key;
+};
+
+const std::vector<test_item> test_list =
```
Comment (evoskuil) [inline test/commands/witness-to-key.cpp]: see above

File: test/generated__find.cpp [hunk 7]
```diff
@@ -163,6 +163,11 @@ BOOST_AUTO_TEST_CASE(generated__find__ec_to_wif__returns_object)
     BOOST_REQUIRE(find("ec-to-wif") != nullptr);
 }
 
+BOOST_AUTO_TEST_CASE(generated__find__ec_to_witness__returns_object)
+{
+    BOOST_REQUIRE(find("ec-to-witness") != nullptr);
+}
+
```
Comment (evoskuil) [inline test/generated__find.cpp]: Generated tests :))

File: src/config/address_format.cpp (hunk 4 above)
Comment (thecodefactory) [reply src/config/address_format.cpp]: I ended up mapping p2wpkh to `witness_pubkey_hash` and p2wsh to `witness_script_hash`.  Hopefully more clear now.

File: src/config/address_format.cpp [hunk 8]
```diff
@@ -86,13 +86,13 @@ std::ostream& operator<<(std::ostream& output, const address_format& argument)
...
     std::string text;
 
-    if (argument.value_ == witness_address::address_format::p2wpkh)
+    if (argument.value_ == witness_address::address_format::witness_pubkey_hash)
     {
-        text = address_format_p2wpkh;
+        text = address_format_witness_pubkey_hash;
     }
-    else if (argument.value_ == witness_address::address_format::p2wsh)
+    else if (argument.value_ == witness_address::address_format::witness_script_hash)
```
Comment (evoskuil) [inline src/config/address_format.cpp]: line length

**Comment by fazlearefin** [comment]
@thecodefactory Is `bx ec-to-witness` part of v3.6.0? I compiled v3.6.0 but the `bx` utility was missing this option

**Comment by thecodefactory** [comment]
@fazlearefin No, this is part of the `master` branch.

**Comment by skaht** [comment]
No **ec-to-witness** shown to the right of https://github.com/libbitcoin/libbitcoin-explorer/wiki 

Not sure if it is worth sharing this deprecated witness-related scripting below.

**m/49’/0’/0’/0/0** BTC private key:

% echo "army van defense carry jealous true garbage claim echo media make crunch" | bx mnemonic-to-seed -p "ionized salt" | bx hd-new -v 76066276 | bx hd-private -d -i 49 | bx hd-private -d -i 0 | bx hd-private -d -i 0 | bx hd-private -i 0 | bx hd-private -i 0 | bx hd-to-ec
134adb604a4055186a7c2dcad2e512d9c0b920fb045c873291968865eaf2ab74

**M/49’/0’/0’/0/0** BTC **`P2WPKH-nested-in-P2SH`** address:

% echo "army van defense carry jealous true garbage claim echo media make crunch" | bx mnemonic-to-seed -p "ionized salt" | bx hd-new -v 76066276 | bx hd-private -d -i 49 | bx hd-private -d -i 0 | bx hd-private -d -i 0 | bx hd-private -i 0 | bx hd-private -i 0 | bx hd-to-ec | bx ec-to-public
02523d3f63f6e0a633355270d5661bd05e1dbf9f692db8791c0215ac2cb4ef2bb1

% echo "army van defense carry jealous true garbage claim echo media make crunch" | bx mnemonic-to-seed -p "ionized salt" | bx hd-new -v 76066276 | bx hd-private -d -i 49 | bx hd-private -d -i 0 | bx hd-private -d -i 0 | bx hd-private -i 0 | bx hd-private -i 0 | bx hd-to-ec | bx ec-to-public | bx bitcoin160
d5858d1e5e0f7f750d32cca66619f3385dc6c15b

% echo "0 [d5858d1e5e0f7f750d32cca66619f3385dc6c15b]" | bx script-encode
0014d5858d1e5e0f7f750d32cca66619f3385dc6c15b

% echo "0 [d5858d1e5e0f7f750d32cca66619f3385dc6c15b]" | bx script-encode | bx sha256 | bx ripemd160 | bx base58check-encode -v 5
382Db9auK4UrFCzWc15Tu8GGCv2RLfVkmc

% echo "0 [d5858d1e5e0f7f750d32cca66619f3385dc6c15b]" | bx script-encode | bx bitcoin160 | bx base58check-encode -v 5
382Db9auK4UrFCzWc15Tu8GGCv2RLfVkmc

% echo "0 [d5858d1e5e0f7f750d32cca66619f3385dc6c15b]" | bx script-to-address -v 5
382Db9auK4UrFCzWc15Tu8GGCv2RLfVkmc

**Pull request merged**